### ISSUE TITLE
Add support for new optimizers

### ISSUE THREAD

**Issue by mnottheone** [PAIR-code/deeplearnjs#58]
Hello everyone ! 

I am willing to contribute to this deeplearn js. I went through roadmap and found adding new optimizers support can be a good starting point for me. 

In order to achieve this, I went through the relevant code and have some doubts. 

Let's suppose I am implementing a new optimizer for SGD with momentum. For this I will need a velocityArrayMap  (or momentumArrayMap) which has to be updated after each batch and then update old variable using it. What I can see in optimizer interface, right now, we have only two TensorArrayMaps for activation and gradients only. So, my doubt is won't we need another TensorArrayMap for velocity ( or momentum) ?

 It was mentioned in roadmap that current optimizer interface is generic enough to support new optimizers like rms , adagrad (both need a cacheVariable ).

 Please help me in figuring out How can I maintain a velocityArrayMap  in cache while training with current interface ?

**Comment by mnottheone** [comment]
Hi !

I realized my doubt was silly, I can define an internal TensorArrayMap in newoptimizer class just like variable gradients and clear them only once when optimizer goes out of scope. 

I have written code for momentumOptimizer which is simple SGD+momentum optimizer and tested it by using in model-builder demo. Seems like, it's working fine, model is getting trained atleast. But I want to check properly i.e it is doing updates as required only. Can anyone suggest how can I do that ?

Also, I am new to open-source contributions but familiar with github, please guide me what all I need to ensure before submitting a pull requests. 

Thanks.

**Comment by dsmilkov** [comment]
Hi! This is exciting! Yes, you would have an internal `TensorArrayMap` which you dispose when the `Optimizer.dispose()` method gets called.

Regarding tests, I agree we are missing an `sgd_optimizer_test.ts` file which can showcase how to test optimizers in general. So that you don't block on us, here is a pointer to a unit test in `session_test.ts` which gives an idea of how to test that the gradient value is correct:
https://github.com/PAIR-code/deeplearnjs/blob/master/src/session_test.ts#L272

You would basically make a session, a simple graph with simple ops, your optimizer, call session.train once (or several times if testing accumulation of momentum) and test that the corresponding entry for the tensor in the gradientArrayMap has the expected value.

Hope this helps. Please let us know if you run into issues :)

**Comment by mnottheone** [comment]
Hi! 
Thanks for all support, my first commit for momentum Optimizer got merged and it feels really great to learn and contribute :) .
Now, I am targeting RMSprop and Adagrad. For that I need sqrt() function for NdArrays, basically which will return ndArray having elementwise sqrt. I can see there is Math.sqrt() but it is on numbers , I don't wanna call it on each element because it will be needed to call get() for all the values in runtime which is not good for speed. 
Other way, it could be handled is by defining math.sqrt(NdArray): NdArray function and using it. Right now, I am not sure how I will do that.

Please correct me if I missed something and help me figuring it out, how it could be done in most efficient manner.  

Thanks

**Comment by nsthorat** [comment]
I will add math.sqrt right now as I actually need it as well!

Just filed #68

**Comment by nsthorat** [comment]
math.sqrt added, you should be good to go.

**Comment by mnottheone** [comment]
Thanks for adding sqrt, seems good to go. Was busy with some other work, will add new optimizers soon.

**Comment by dsmilkov** [comment]
Closing this thread, since we already have 7 optimizers with the [0.5.0 release](https://github.com/PAIR-code/deeplearnjs/releases/tag/v0.5.0). See docs [here](https://deeplearnjs.org/docs/api/index.html#Training-Optimizers).
